cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -0,0 +1,419 @@
+
+    // instantiate a dataproc job controller client
+    CredentialsProvider credentialsProvider = FixedCredentialsProvider.create(credentials);
+    this.jobControllerClient = JobControllerClient.create(
+      JobControllerSettings.newBuilder().setCredentialsProvider(credentialsProvider)
+        .setEndpoint(region + DATAPROC_GOOGLEAPIS).build());
+  }
+
+  @Override
+  public RuntimeJobId launch(RuntimeJobInfo runtimeJobInfo) throws Exception {
+    ProgramRunInfo runInfo = runtimeJobInfo.getProgramRunInfo();
+    LOG.info("Starting to launch program run {} with following configurations: cluster {}, project {}, region {}, " +
+               "bucket {}.", runInfo.getRun(), clusterName, projectId, region, bucket);
+
+    File tempDir = Files.createTempDirectory("dataproc.launcher").toFile();
+    // on dataproc bucket the run root will be <bucket>/cdap-job/<runid>/. All the files for this run will be copied
+    // under that base dir.
+    String runRootPath = getPath(CDAP_GCS_ROOT, runInfo.getRun());
+    try {
+      // step 1: build twill.jar and launcher.jar and add them to files to be copied to gcs
+      List<RuntimeLocalFile> localFiles = getRuntimeLocalFiles(runtimeJobInfo.getLocalizeFiles(), tempDir);
+
+      // step 2: upload all the necessary files to gcs so that those files are available to dataproc job
+      for (RuntimeLocalFile fileToUpload : localFiles) {
+        String targetFilePath = getPath(runRootPath, fileToUpload.getName());
+        LOG.debug("Uploading file {} to gcs bucket {}.", targetFilePath, bucket);
+        uploadFile(targetFilePath, fileToUpload);
+        LOG.info("Uploaded file {} to gcs bucket {}.", targetFilePath, bucket);
+      }
+
+      // step 3: build the hadoop job request to be submitted to dataproc
+      SubmitJobRequest request = getSubmitJobRequest(runtimeJobInfo.getRuntimeJobClass().getName(),
+                                                     runInfo, localFiles);
+
+      // step 4: submit hadoop job to dataproc
+      LOG.info("Submitting hadoop job {} to cluster {}.", request.getJob().getReference().getJobId(), clusterName);
+      Job job = jobControllerClient.submitJob(request);
+      LOG.info("Successfully submitted hadoop job {} to cluster {}.", job.getReference().getJobId(), clusterName);
+
+      return new RuntimeJobId(job.getReference().getJobId());
+    } catch (Exception e) {
+      // delete all uploaded gcs files in case of exception
+      deleteGCSPath(runRootPath);
+      throw new Exception(String.format("Error while launching job %s on cluster %s",
+                                        runInfo.getRun(), clusterName), e);
+    } finally {
+      // delete local temp directory
+      deleteDirectoryContents(tempDir);
+    }
+  }
+
+  @Override
+  public Optional<RuntimeJobDetail> getDetail(RuntimeJobId runtimeJobId) throws Exception {
+    String jobId = runtimeJobId.getRuntimeJobId();
+
+    Optional<RuntimeJobDetail> jobDetails = Optional.empty();
```

Comment:
No need to have this. Just return in place.

try {
  ...
  return Optional.of(new RuntimeJobDetail(...));
} catch (ApiException e) {
  ...
}
return Optional.empty();